src/shell/shell_cmds.cpp

```diff
@@ -2532,6 +2492,98 @@ static std::vector<std::string> search_files(const std::string_view query)
 	return files;
 }
 
+struct FileExt {
+	std::string root      = {};
+	std::string extension = {};
+};
+
+static FileExt split_extention(const std::string& fullname)
+{
+	FileExt ret = {};
+	size_t pos  = fullname.rfind('.');
+	ret.root    = fullname.substr(0, pos);
+	++pos;
+	if (pos > 0 && pos < fullname.size()) {
+		ret.extension = fullname.substr(pos);
+	} else {
+		ret.extension = {};
+	}
+	return ret;
+}
+
+static std::string handle_wildcards(const std::string& wild, const std::string& old)
+{
+	std::string ret = {};
+	for (size_t i = 0; i < wild.size(); ++i) {
+		char c    = wild[i];
+		bool done = false;
+		switch (c) {
+		case '*':
+			if (i < old.size()) {
+				ret.append(old, i);
+			}
+			done = true;
+			break;
+		case '?':
+			if (i < old.size()) {
+				ret.push_back(old[i]);
+			}
+			break;
+		default: ret.push_back(c);
+		}
+		if (done) {
+			break;
+		}
+	}
+	return ret;
+}
+
+void DOS_Shell::CMD_RENAME(char* args)
+{
+	HELP("RENAME");
+
+	const std::string first_arg  = DOS_Canonicalize(strip_word(args));
+	const std::string second_arg = strip_word(args);
+	if (first_arg.empty() || second_arg.empty()) {
+		SyntaxError();
+		return;
+	}
+
+	// Second argument must not contain a path
+	if (second_arg.find_first_of("\\:") != std::string::npos) {
+		SyntaxError();
+		return;
+	}
+
+	const std::string path = first_arg.substr(0, first_arg.rfind('\\') + 1);
+	const FileExt wild     = split_extention(second_arg);
+
+	// Search for files matching the first argument (may be multiple files
+	// due to wildcards)
+	for (const std::string& old_filename : search_files(first_arg)) {
+		const FileExt old_split = split_extention(old_filename);
+
+		FileExt new_split = {};
+		new_split.root    = handle_wildcards(wild.root, old_split.root);
+		new_split.extension = handle_wildcards(wild.extension,
+		                                       old_split.extension);
+
+		std::string old_fullpath = path + old_filename;
+
+		std::string new_fullpath = path + new_split.root;
+		if (!new_split.extension.empty()) {
+			new_fullpath.push_back('.');
+			new_fullpath.append(new_split.extension);
+		}
+
+		if (!DOS_Rename(old_fullpath.c_str(), new_fullpath.c_str())) {
+			WriteOut("Rename %s -> %s failed\n",
```

Comment:
Question (not asking for changes, not sure of the answer) --
In real DOS, I wonder if a failed rename sets `ERRORLEVEL` that a subsequent batch statement can check? (or something like that?) - maybe that happens at another level though (the filesystem calls?).

Reply:
`DOS_SetError` gets called inside `DOS_Rename` if it fails.